### PR TITLE
Samarth agarwal1/pypackage

### PR DESCRIPTION
**Pull request by Samarth-Agarwal1**
To build it, run `python setup.py sdist bdist_wheel` at the root level. To install the package go into the `dist` folder and run `pip install .\test_data-0.0.0-py3-none-any.whl`. That will only work if you have `wheel` installed in your pip library so make sure you have it installed prior to trying to run that install command. Now if you run `test_data` in the command line you should be able to run the distribution. You can also run it by going into Python console and doing `import test_data.generate_vcf` and then running `test_data.generate_vcf.main()`. In `setup.py` I am not sure if the parameter `entry_points` for `setuptools.setup()` needs to be changed or could be better.

### CODE REVIEW

**Comment by Samarth-Agarwal1** [comment]
Also, there should probably be a license for this before this is merged with master.

**Comment by Samarth-Agarwal1** [comment]
I am also not sure if there should be more parameters passed in to setuptools.setup() for our purposes or if any of the other parameters should be changed.

**Review by amanjeev — APPROVED** [review]
LGTM! I was able to create a `virtualenv` and `pip install -e .` this package. 

```
test_data on  Samarth-Agarwal1/pypackage [?] 
10:39:57 ❤ pip install -e .                                                                                                            
Obtaining file:///home/aj/proj/candig/test_data
Collecting docopt==0.6.2 (from test-data==0.0.0)
  Using cached https://files.pythonhosted.org/packages/a2/55/8f8cab2afd404cf578136ef2cc5dfb50baa1761b68c9da1fb1e4eed343c9/docopt-0.6.2.tar.gz
Collecting numpy==1.19.0 (from test-data==0.0.0)
  Using cached https://files.pythonhosted.org/packages/57/55/5366b0afed57facaf887b05f0e98c887a9e27021e3f4e78a26a6d535e538/numpy-1.19.0-cp38-cp38-manylinux2010_x86_64.whl
Installing collected packages: docopt, numpy, test-data
  Running setup.py install for docopt ... done
  Running setup.py develop for test-data
Successfully installed docopt-0.6.2 numpy-1.19.0 test-data
WARNING: You are using pip version 19.2.3, however version 20.1.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.



test_data on  Samarth-Agarwal1/pypackage [?] took 3s 
10:40:06 ❤ pip list                                                                                                                    
Package    Version Location                          
---------- ------- ----------------------------------
docopt     0.6.2   
numpy      1.19.0  
pip        19.2.3  
setuptools 41.2.0  
test-data  0.0.0   /home/aj/proj/candig/test_data/src
WARNING: You are using pip version 19.2.3, however version 20.1.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.

test_data on  Samarth-Agarwal1/pypackage [?] took 316ms 
10:40:11 ❤ test_data                                                                                                                   
Usage:
  generate_vcf.py <number_of_files> <output_folder> <filename_start>
```

**Comment by Samarth-Agarwal1** [comment]
One thing I just noticed: When you run `test_data` with the wrong/incorrect command-line parameters, the program outputs to the console:
```
Usage:
  generate_vcf.py <number_of_files> <output_folder> <filename_start>
```

Instead of showing `generate_vcf.py`, shouldn't it show the name of the executable so that the user knows that you have to run it with the executable and then the command-line arguments; you're not supposed to run it by writing `generate_vcf.py`.

**Comment by ljdursi** [comment]
Yes, now that you have a package it gets installed as test_data (although e.g. python3 .src/test_data/generate_vcf.py would also work).  This uses docops for help and you can edit the message by editing the module documentation (comments at the head).

I think one thing this would need before being a successful package would be having some way to download the dbSNP file that Zoltan used for this.  You might be able to get the file itself from @haoyuanli and we could put it somewhere (maybe abusing the releases functionality of github?) or we could have a script for downloading the full dbsnp file and filtering it.

@haoyuanli , @amanjeev , what do you think - does the dbsnp file handling go in this PR or can this be merged (after fixing the help message) and the data dependency can go in the next one?

**Comment by Samarth-Agarwal1** [comment]
Is there a way to dynamically change the Usage message depending on whether the user is running it as
 `python src/test_data/generate_vcf.py` or `test_data`?

If the user runs `.src/test_data/generate_vcf.py` the output is:
```
Usage:
  generate_vcf.py <number_of_files> <output_folder> <filename_start>
```

If the user runs it as `test_data` the output is:
```
Usage:
  test_data <number_of_files> <output_folder> <filename_start>
```

**Comment by ljdursi** [comment]
Probably?  But there's nothing wrong with having a recommended way to run it, and if you've built a package for this there's no reason why the recommended usage shouldn't be to use the package name.

**Comment by amanjeev** [comment]
> @haoyuanli , @amanjeev , what do you think - does the dbsnp file handling go in this PR or can this be merged (after fixing the help message) and the data dependency can go in the next one?

 @haoyuanli what are your thoughts? I am fine either way but inclined towards merging this as-is.

**Comment by amanjeev** [comment]
@Samarth-Agarwal1 : you're asking good questions! :slightly_smiling_face:

**Comment by jimmyhli** [comment]
> One thing I just noticed: When you run `test_data` with the wrong/incorrect command-line parameters, the program outputs to the console:
> 
> ```
> Usage:
>   generate_vcf.py <number_of_files> <output_folder> <filename_start>
> ```
> 
> Instead of showing `generate_vcf.py`, shouldn't it show the name of the executable so that the user knows that you have to run it with the executable and then the command-line arguments; you're not supposed to run it by writing `generate_vcf.py`.

Yeah, but since you are defining a command (console script) that points to this script already, you can just document the usage of this preset command, and there's no need to "dynamically change the Usage message", just ask users to run the preset command.

I am a little uncertain about the name of this command, however, `test_data` is a little ambiguous

**Comment by Samarth-Agarwal1** [comment]
I named it `test_data` because that is the name of the repo; I agree that it isn't a very descriptive name but the repo name will be need to be changed later on anyways as indicated in the Google docs. When the repo name is changed the `setup.py` will probably be need to be refactored to accomodate the new repo change and the command will likely be changed to match the new repo name.

**Comment by jimmyhli** [comment]
> I named it `test_data` because that is the name of the repo; I agree that it isn't a very descriptive name but the repo name will be need to be changed later on anyways as indicated in the Google docs. When the repo name is changed the `setup.py` will probably be need to be refactored to accomodate the new repo change and the command will likely be changed to match the new repo name.

Do you have a list of names in mind, as potential candidates? If there's a name you have in mind already, we can discuss it here.

Other than that, I think the new command should be documented, probably just mention it in the README.md

**Comment by Samarth-Agarwal1** [comment]
I was thinking `generate_sample_data` as the repo name. It's short and clearly states what the purpose of the repo is.

**Comment by jimmyhli** [comment]
> I was thinking `generate_sample_data` as the repo name. It's short and clearly states what the purpose of the repo is.

That sounds fine to me. It's up to you whether you want to address that in a separate PR; for the scope of this PR, I think it would be good to document the newly-added command to the README.md, and just briefly describe what it does. 

~Can you remind me what `numpy` does here? Where is it imported?~ Nvm

**Comment by ljdursi** [comment]
I think with the new repo name you can change the package name, README, and module docstring (and so the name that shows up in the usage) and you should be good to go

**Comment by Samarth-Agarwal1** [comment]
I just pushed some more commits which modify the README and refactor some code to make it compatible with the new repo name.

File: README.md
Comment (jimmyhli) [inline README.md]: a space is needed after `#` for markdown to recognize this as a title, same applies to all of your titles below, including the `##` ones.

**Review by jimmyhli — CHANGES_REQUESTED** [review]

File: README.md
Comment (jimmyhli) [inline README.md]: same here

File: README.md
Comment (jimmyhli) [inline README.md]: same here

**Comment by Samarth-Agarwal1** [comment]
I've added a space after every # and ## in the README.

**Comment by jimmyhli** [comment]
> I've added a space after every # and ## in the README.

Thanks. Besides I will need to look up where this sample dbSNP file is, I will let u know once i find it

File: README.md
Comment (jimmyhli) [inline README.md]: @amanjeev has mentioned in his earlier comments how he installed the package with `pip install -e .`, i think this should be documented, and possibly as a preferred way for local development, it's just easier since it doesn't require the installation of `wheel`.

**Comment by amanjeev** [comment]
@Samarth-Agarwal1 a good start to learn a bit more about Markdown is perhaps GH's tutorial: https://guides.github.com/features/mastering-markdown/. This also serves as my own cheatsheet. 

I would not have started at GH but given that we mostly use Markdown on GH, it is ok. Also, luckily they have a section "GitHub Flavored Markdown".

File: README.md
Comment (Samarth-Agarwal1) [reply README.md]: What exactly does the `-e` flag do? I know it means editable mode. but how is it different from just running `pip install` without the `-e` flag?

Comment (jimmyhli) [reply README.md]: @amanjeev can correct me, but `-e` is indeed the editable mode which gives you an editable installation for dev use, i believe `pip install -e .` is largely equivalent to `python setup.py develop`, while `pip install .` is equivalent to `python setup.py install`; so yeah the `-e` is more meant for development here i guess

Comment (jimmyhli) [reply README.md]: see https://stackoverflow.com/questions/19048732/python-setup-py-develop-vs-install for a more complete answer

Comment (amanjeev) [reply README.md]: Sorry Jimmy, I misread your comment. Yes, you are correct. :)

Comment (Samarth-Agarwal1) [reply README.md]: I am still a little confused. For the documentation, wouldn't it make more sense to write `pip install .` since we aren't expecting the user to do any development?

Comment (amanjeev) [reply README.md]: Good question! I think in usage, we should do `pip install package` and under instructions for development we can say how to install as editable, how to test etc. What do you think of this?

Comment (Samarth-Agarwal1) [reply README.md]: Yeah, I think that is a good idea. I don't have a Development section however. Where should I add that section?

**Comment by Samarth-Agarwal1** [comment]
I have added detail on how to install the program from source code. I have created a heading called "Development" at the bottom but I'm not sure what to write for it.

**Comment by amanjeev** [comment]
> I have created a heading called "Development" at the bottom but I'm not sure what to write for it.

@Samarth-Agarwal1 : This is a great question and like everything else in engineering, it requires a trade-off. 

Think of it this way - what would you tell someone new to this repo on where to start if they wanted to contribute? Say someone is absolutely new and just want to have stepwise instructions on what they have to do. 

This can go to as extremely detailed as starting with "What is Python?" but we likely do not want that. Another extreme would be that we do not say anything about development but we likely do not want that either. To me, a balanced approach would be starting with the commands on how to create a `virtualenv` and activate it, how to install this package as _editable_ in it, how to run tests and add tests etc.

We likely want to give guidelines about commit messages standards and code-of-conduct but that work is being done separately.

File: README.md
Comment (amanjeev) [inline README.md]: Since, we already show the `-h` help below, do you think this section adds value to the documentation?

Comment (Samarth-Agarwal1) [reply README.md]: Not really, I guess it is kind of redundant.

File: README.md
Comment (amanjeev) [inline README.md]: I wonder if this line is redundant since the command is covered already. What do you think?

File: src/sample_data_generator/generate_vcf.py
Comment (amanjeev) [inline src/sample_data_generator/generate_vcf.py:7]: yay! :slightly_smiling_face:

**Review by amanjeev — CHANGES_REQUESTED** [review]

**Review by amanjeev — APPROVED** [review]

**Review by ljdursi — APPROVED** [review]
Looks good!

**Comment by Samarth-Agarwal1** [comment]
I added some details for the Development section. Should I add anything more to the Development section or change it up?

**Comment by Samarth-Agarwal1** [comment]
Also, in order for this project to work the user needs to have Python 3.7. Where should I mention in the README that they need a minimum version of Python 3.7?

**Comment by amanjeev** [comment]
> Also, in order for this project to work the user needs to have Python 3.7. Where should I mention in the README that they need a minimum version of Python 3.7?

[It is a good idea but I think we can just mention that in `setup.py`, which you've already done](https://github.com/CanDIG/sample-data-generator/blob/Samarth-Agarwal1/pypackage/setup.py#L22)  :slightly_smiling_face: 

To clarify further (sorry I should have done this earlier), this is my personal preference to use the default, de facto places to find certain attributes of our things. There is no harm in mention the Python version in the README and in fact might be helpful. I am mostly concerned with any non-automated updates that one has to do as a programmer. Either way, it is your call.

**Review by jimmyhli — APPROVED** [review]

**Pull request merged**
